wordpress-mobile/WordPress-Android
# Refactor concurrent uploading of local changes

Resolves #17463

This PR adapts the logic for uploading local changes to the cloud and introduces unit tests for the batch job of uploading all local sites posts & pages.

We've [noticed in Sentry](https://a8c.sentry.io/discover/results/?name=CompletionHandlerException&project=5731682&query=issue:JETPACK-ANDROID-1VC&sort=-release&statsPeriod=90d) the crash still occurs, and attempted to fix it via the `mutex.withLock { … }` workaround that we've merged via PR #18067 into `22.4`. Unfortunately it didn't eradicate the crash as seen in [this Sentry report](https://a8c.sentry.io/discover/results/?environment=debug&environment=prod&environment=release&field=title&field=event.type&field=project&field=user.display&field=timestamp&field=replayId&name=All+Events&project=5731682&query=release%3A22.4-rc-1+title%3A%22CompletionHandlerException%3A+Exception+in+resume+onCancellation+handler+for+CancellableContinuation%28DispatchedContinuation%5BDis...%22&sort=-timestamp&statsPeriod=90d&yAxis=count%28%29) for `22.4-rc-1`.

> **Note** To reviewers
> Would love your input on steps 2 & 3 from below. Please don't merge, I will handle the merge after collecting more input.

## To Test

1. **Verify** local draft posts and pages are uploaded
2. **Review** thoroughly the code changes and [this unit test](https://github.com/wordpress-mobile/WordPress-Android/pull/18449/files#diff-ac70facd6149acaa6c5c582c70cdcb36b7150a4422586cc0bfe291de429258d3R215-R235)
3. **Brainstorm** what other tests / ideas I could try
4. **Verify** CI unit tests job succeeded

## Regression Notes
1. Potential unintended areas of impact
   Offline posts and pages upload

7. What I did to test those areas of impact (or what existing automated tests I relied on)
   Manual testing and unit testing

8. What automated tests I added (or what prevented me from doing so)
   Unit tests for the updated code

PR submission checklist:

- [x] UI Changes testing checklist: n/a
